sdk/servicebus/Microsoft.Azure.ServiceBus/src/MessageReceivePump.cs

```diff
@@ -206,77 +228,194 @@ void CancelAutoRenewLock(object state)
                     // by the renewLockCancellationToken. This way we prevent a TaskCanceledException.
-                    var delayTask = await Task.Delay(amount, renewLockCancellationToken)
+                    var delayTask = await Task.Delay(renewAfterTimeSpan, renewLockCancellationToken)
                         .ContinueWith(t => t, TaskContinuationOptions.ExecuteSynchronously)
                         .ConfigureAwait(false);
                     if (delayTask.IsCanceled)
                     {
                         break;
                     }
 
-                    if (!this.pumpCancellationToken.IsCancellationRequested &&
-                        !renewLockCancellationToken.IsCancellationRequested)
+
+                    bool batchRenewalSucceeded;
+
+                    if (renewableMessages.Count > 1)
                     {
-                        await this.messageReceiver.RenewLockAsync(message).ConfigureAwait(false);
-                        MessagingEventSource.Log.MessageReceiverPumpRenewMessageStop(this.messageReceiver.ClientId, message);
+                        try
+                        {
+                            await this.messageReceiver.RenewLockAsync(renewableMessages).ConfigureAwait(false);
+
+                            batchRenewalSucceeded = true;
+                        }
+                        catch
+                        {
+                            batchRenewalSucceeded = false;
+                        }
                     }
                     else
                     {
-                        break;
+                        batchRenewalSucceeded = false;
                     }
+
+                    if (!batchRenewalSucceeded)
+                    {
+                        IList<Message> noLongerRenewableMessages = null;
+
+                        foreach (var renewableMessage in renewableMessages)
+                        {
+                            try
+                            {
+                                if (!this.pumpCancellationToken.IsCancellationRequested &&
+                                    !renewLockCancellationToken.IsCancellationRequested)
+                                {
+                                    await this.messageReceiver.RenewLockAsync(renewableMessage).ConfigureAwait(false);
+                                }
+                                else
+                                {
+                                    break;
+                                }
+                            }
+                            catch (Exception exception)
+                            {
+                                MessagingEventSource.Log.MessageReceiverPumpRenewMessageException(this.messageReceiver.ClientId, new[] { renewableMessage }, exception);
+                                await this.RaiseExceptionReceived(exception, new[] { renewableMessage }, ExceptionReceivedEventArgsAction.RenewLock).ConfigureAwait(false);
```

Comment:
Same as `AbandonMessagesIfNeededAsync` but even worse as this is a time-sensitive operation.